_posts/2020-09-06-how-the-system-tests-works.md

```diff
@@ -0,0 +1,603 @@
+apiVersion: kafka.strimzi.io/v1beta1
+kind: KafkaTopic
+metadata:
+  name: my-topic
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  partitions: 1
+  replicas: 1
+  config:
+    retention.ms: 7200000
+    segment.bytes: 1073741824
+```
+
+##### <a id="kafkaimplementation">Code snippet 2.2 Loading of the YAML file to JAVA representation object.</a>
+```java
+KafkaTopic kafkaTopic = getKafkaTopicFromYaml(PATH_TO_KAFKA_TOPIC_CONFIG); 
+// additional changes...
+kafkaTopic = new KafkaTopicBuilder(kafkaTopic)
+    .withNewMetadata()
+        .withName(topicName)
+        .withNamespace(ResourceManager.kubeClient().getNamespace())
+        .addToLabels(Labels.STRIMZI_CLUSTER_LABEL, clusterName)
+    .endMetadata()
+    .editSpec()
+        .withPartitions(partitions)
+        .withReplicas(replicas)
+        .addToConfig("min.insync.replicas", minIsr)
+    .endSpec();
+```
+
+Everything is built within three stacks. These three stacks are managed by `ResourceManager`, which is taking care of switching. 
+First one's, `method stack`, responsibility is to hold all resources, which are invoked in the test cases. 
+Second one's, the `class stack`, responsibility is to hold all resources, which are invoked in the `@BeforeAll` notation. 
+The third stack is used as a pointer to either class or method stack to delete proper resources in specific test phases.
+The `pointer stack` points to the `class stack` in we are in the `@BeforeAll` scope, otherwise we are working with the `method stack`. 
+The logic inside these stacks are that once you create some resource (for instance KafkaTopic),
+it will be pushed inside the stack and the deletion of topic will be performed in the teardown phase. 
+Once the teardown phase is initiated, the `method stack` will pop all resources after test case is over. The same logic applies with `class stack,
+which will pop all resources when test suite is over. 
+What is worth to mention, is that if you specify `SKIP_TEARDOWN` environment variable to `TRUE` it will skip the teardown 
+phase and all created resources will remain in the stack.
+
+## Lifecycle of tests
+
+Like any other test life cycle, we follow the [SEVT](http://xunitpatterns.com/Four%20Phase%20Test.html) convention. 
+[SEVT](http://xunitpatterns.com/Four%20Phase%20Test.html) stands for Setup, Exercise, Verify and Teardown.
+The fundamental idea can be illustrated as follows:
+
+```
+1.SETUP        | instance = new Instance(); 
+2.EXERCISE     | instance.setSomeValue(someValue);
+3.VERIFY       | asserThat(instance.getSomeValue(), is(someValue));
+4.TEARDOWN    | instance = null; (this steps is prevailingly done by testing Frameworks)
+```
+
+#### Setup
+
+In `Setup phase` we perform following actions:
+
+* Create namespace(s)
+* Deploy the Strimzi Cluster operator
+* Deploy the Kafka cluster and/or other components (optional)
+
+The reason why the last point is optional, is because we have some test cases where we might want to have a different kafka 
```

Comment:
Following the table above, I'm not sure this is true because from there it does not sound like setup is only the `@BeforeClass`. It could be also the first part of the test method.